vimscript.html.markdown

```diff
@@ -0,0 +1,572 @@
+
+### Function namespaces
+
+See `:h write-library-script`, `:h autoload`
+
+```vim
+" Must be defined in autoload/foo/bar.vim
+function! foo#bar#log(value)
+    echomsg value
+endfunction
+
+call foo#bar#log('Hello')
+```
+
+
+## Frequently used ex-commands
+
+### Sourcing runtime files
+
+See `:h 'runtimepath'`
+
+```vim
+" Source first match among runtime paths
+runtime plugin/my-plugin.vim
+```
+
+### Defining new ex-commands
+See `:h 40.2`, `:h :command`
+
+```vim
+" First argument is the name of the command, rest is the command body
+command! SwapLines normal ddp
+```
+
+The exclamation mark works the same as with `:function`. User-defined commands
+must start with a capital letter. The `:command` command can take many extra
+optional arguments, such as `-nargs`, all of them start with a dash.
+
+### Defining auto-commands
+See `:h 40.3`, `:h autocmd`
+
+```vim
+" Events, patterns, commands
+autocmd Filetype html,xml :set tabstop=2
+```
+
+Events and patterns are separated by commas with no space between. See `:h
+autocmd-events` for standard events.
+
+#### Auto groups
+
+When a file is source multiple times the auto-commands are defined anew,
+without deleting the old ones, causing auto-commands to pile up over time. Use
+auto-groups and the following ritual to guard against this.
+
+```vim
+augroup sgml-like   |" The name of the group is arbitrary
+		autocmd!        |" Deletes all auto-commands in the current group
+		autocmd Filetype html,xml :set tabstop=2
+augroup END         |" Switch back to default auto-group
+```
+
+### Executing (run-time macros of sorts)
+
+Sometimes we need to construct an ex-command where part of the command is not
+known until runtime.
+
+```vim
+let line = 3               |" Line number deternimed at runtime
+execute line . 'delete'    |" Delete a line
+```
+
+### Executing normal-mode commands
+
+Use `:normal` to play back a sequence of normal mode commands from the
+command-line. Add an exclamation mark to ignore user mappings.
+
+```vim
+normal! ggddGp             |" Transplant first line to end of buffer
+
+" Window commands can be used with :normal, or with :wincmd if :normal would
+" not work
+wincmd L                   |" Move current window all the way to the right
+```
+
+
+## Frequently used functions
+
+```vim
+" Feature check
+echo  has('nvim')                  |" Running Neovim
+echo  has('python3')               |" Support for Python 3 plugins
+echo  has('unix')                  |" Running on a Unix system
+echo  has('win32')                 |" Running on a Windows system
+
+
+" Test if something exists
+echo  exists('g:my_var')           |" Variable (scope optional)
+echo  exists('MyFunc')             |" Function
+echo  exists('my_dict["foo"]')     |" Dictionary entry
+echo  exists('my_dict["foo"]')     |" Dictionary entry
+echo  exists('$TERM')              |" Environment variable
+echo  exists(':echo')              |" Command
+echo  exists('+tui')               |" Feature
```

Comment:
Adding more examples here would be beneficial. This set of examples from `:h exists()` is more comprehensive. (Also line 54[12] are duplicates).

```vim
echo exists("&mouse")                     |" Option exists
echo exists("+mouse")                     |" Option exists && works
echo exists("$HOSTNAME")                  |" Environment variable
echo exists("*strftime")                  |" Built-in function
echo exists("*s:MyFunc")                  |" User defined function
echo exists("bufcount")                   |" Variable
echo exists(":Make")                      |" Command
echo exists("#CursorHold")                |" autocmd for this event in the current augroup
echo exists("#BufReadPre#*.gz")           |" autocmd for this event and pattern
echo exists("#filetypeindent")            |" augroup
echo exists("#filetypeindent#FileType")   |" augroup and autocmd for this event
echo exists("#filetypeindent#FileType#*") |" augroup and autocmd for this event and pattern
echo exists("##ColorScheme")              |" autocmd for this event is supported
```

Reply:
Also, line 540 isn't entirely accurate.

Reply:
Fixed.